node/network/collator-protocol/src/validator_side.rs

```diff
@@ -0,0 +1,1369 @@
+	para_id: ParaId,
+	mut tx: mpsc::Sender<(CollatorId, CandidateReceipt, PoV)>
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	// First take a look if we have already stored some of the relevant collations.
+	if let Some(collations) = state.collations.get(&(relay_parent, para_id)) {
+		for collation in collations.iter() {
+			if let Err(e) = tx.send(collation.clone()).await {
+				// We do not want this to be fatal because the receving subsystem
+				// may have closed the results channel for some reason.
+				trace!(
+					target: TARGET,
+					"Failed to send collation: {:?}", e,
+				);
+			}
+		}
+	}
+
+	// Dodge multiple references to `state`.
+	let mut relevant_advertisers = Vec::new();
+
+	// Who has advertised a relevant collation?
+	for (k, v) in state.advertisments.iter() {
+		if v.contains(&(para_id, relay_parent)) {
+			relevant_advertisers.push(k.clone());
+		}
+	}
+
+	// Request the collation.
+	// Assume it is `request_collation`'s job to check and ignore duplicate requests.
+	for peer_id in relevant_advertisers.into_iter() {
+		request_collation(ctx, state, relay_parent, para_id, peer_id, Some(tx.clone())).await?;
+	}
+
+	Ok(())
+}
+
+/// Report a collator for some malicious actions.
+async fn report_collator<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	id: CollatorId,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	// Since we have a one way map of PeerId -> CollatorId we have to
+	// iterate here. Since a huge amount of peers is not expected this
+	// is a tolerable thing to do.
+	for (k, v) in state.known_collators.iter() {
+		if *v == id {
+			modify_reputation(ctx, k.clone(), COST_REPORT_BAD).await?;
+		}
+	}
+
+	Ok(())
+}
+
+/// Some other subsystem has reported a collator as a good one, bump reputation.
+async fn note_good_collation<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	id: CollatorId,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	for (peer_id, collator_id) in state.known_collators.iter() {
+		if id == *collator_id {
+			modify_reputation(ctx, peer_id.clone(), BENEFIT_NOTIFY_GOOD).await?;
+		}
+	}
+
+	Ok(())
+}
+
+/// A peer's view has changed. A number of things should be done:
+///  - Ongoing collation requests have to be cancelled.
+///  - Advertisments by this peer that are no longer relevant have to be removed.
+async fn handle_peer_view_change(
+	state: &mut State,
+	peer_id: PeerId,
+	view: View,
+) -> Result<()> {
+	let current = state.peer_views.entry(peer_id.clone()).or_default();
+
+	let removed: Vec<_> = current.difference(&view).cloned().collect();
+
+	*current = view;
+
+	if let Some(advertisments) = state.advertisments.get_mut(&peer_id) {
+		advertisments.retain(|(_, relay_parent)| !removed.contains(relay_parent));
+	}
+
+	let mut requests_to_cancel = Vec::new();
+
+	for removed in removed.into_iter() {
+		state.requested_collations.retain(|k, v| {
```

Comment:
`O(m*n)` where `m` is `removed.len()` and `n` is `state.requested_collations.len()`.

We could reduce that algorithmic complexity if we require that both collections be sorted, and then do a double-pointer merge traversal, but I don't have a sense of the sizes involved here. Is it even worth worrying about that?

Reply:
I highly doubt that this number ever going to be huge enough to worry about complexity here. Even if we have a para with like 500 collators it is unlikely that a huge number of those will come up with blocks at each point in time because consensus reasons, is it?

Reply:
I am fine with it as-is for now. The iteration should be fast on good hardware.